mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,300 @@
+    MlemBuilder,
+    MlemDeployment,
+    MlemEnv,
+    MlemModel,
+)
+from mlem.runtime.client import Client, HTTPClient
+from mlem.runtime.server import Server
+from mlem.ui import EMOJI_OK, echo
+
+from ..docker.base import (
+    DockerDaemon,
+    DockerImage,
+    DockerRegistry,
+    generate_docker_container_name,
+)
+from .build import build_k8s_docker
+from .context import K8sYamlBuildArgs, K8sYamlGenerator
+from .utils import create_k8s_resources, namespace_deleted, pod_is_running
+
+POD_STATE_MAPPING = {
+    "Pending": DeployStatus.STARTING,
+    "Running": DeployStatus.RUNNING,
+    "Succeeded": DeployStatus.STOPPED,
+    "Failed": DeployStatus.CRASHED,
+    "Unknown": DeployStatus.UNKNOWN,
+}
+
+
+def find_index(nodes_list, node_name):
+    for i, each_node in enumerate(nodes_list):
+        if each_node.metadata.name == node_name:
+            return i
+    return -1
+
+
+class ServiceType(MlemABC):
+
+    abs_name: ClassVar = "k8s_service_type"
+    type: ClassVar = "kubernetes"
+
+    class Config:
+        type_root = True
+
+    def get_host_and_port(
+        self, service, namespace="mlem"  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[str], Optional[int]]:
+        """Returns host and port for the service in Kubernetes"""
+        return None, None
+
+
+class NodePortService(ServiceType):
+
+    type: ClassVar = "nodeport"
+
+    def get_host_and_port(self, service, namespace="mlem"):
+        host, port = None, None
+        port = service.items[0].spec.ports[0].node_port
+        node_name = (
+            client.CoreV1Api()
+            .list_namespaced_pod(namespace)
+            .items[0]
+            .spec.node_name
+        )
+        node_list = client.CoreV1Api().list_node().items
+        node_index = find_index(node_list, node_name)
+        if node_index == -1:
+            raise DeploymentError(
+                f"Couldn't find the node where pods in namespace {namespace} exists"
+            )
+        address_dict = node_list[node_index].status.addresses
+        external_ip_found = False
+        for each_address in address_dict:
+            if each_address.type == "ExternalIP":
+                host = each_address.address
+                external_ip_found = True
+        if not external_ip_found:
+            raise EndpointNotFound(
+                f"Node {node_name} doesn't have an externally reachable IP address"
+            )
+        return host, port
+
+
+class LoadBalancerService(ServiceType):
+
+    type: ClassVar = "loadbalancer"
+
+    def get_host_and_port(self, service, namespace="mlem"):
+        host, port = None, None
+        port = service.items[0].spec.ports[0].port
+        ingress = service.items[0].status.load_balancer.ingress[0]
+        host = ingress.hostname or ingress.ip
+        return host, port
+
+
+class ClusterIPService(ServiceType):
+
+    type: ClassVar = "clusterip"
+
+    def get_host_and_port(self, service, namespace="mlem"):
+        raise TypeError(
```

Comment:
Why type error?

Reply:
Also all exceptions not subclassed from MlemError are treated as unexpected. So this should be MlemError

Reply:
okay, thanks, will do

Reply:
done